### DISCUSSION TITLE
Page Forms Examples

### DISCUSSION THREAD

**Discussion by tholzheim** [SmartDataAnalytics/OpenResearch#108] [Ideas]
As proposed in #104 I have looked into nearly all pages listed on https://wikiapiary.com/wiki/Extension:Page_Forms
Most just use the basic functionality of the [Page Forms Extensions](https://rq.bitplan.com/index.php/Page_Forms) https://www.mediawiki.org/wiki/Extension:Page_Forms 
One of the most advanced uses of page forms is this one https://directory.fsf.org/wiki?title=Blender&action=formedit
![image](https://user-images.githubusercontent.com/45004115/108325406-55fde580-71c9-11eb-8831-4c335254dfe5.png)
So basically the current OPENRESEARCH is state of the art. Further improvements can only be done with technical tricks that might harm the maintainability of the site.
@MusaabKh and I are currently working on a prototype see https://confident.dbis.rwth-aachen.de/wikicms/

**Comment by WolfgangFahl** [comment]
https://fuutxr.axshare.com/#g=1&p=enter_event is inspired by [Zenodo ](https://zenodo.org/)which is based on a  totally different platform and backed by a much larger development team. I think we should not try to imitate this if leads to unclean technical solutions and extra effort. The http://wiki.bitplan.com/index.php/Template:HideShow approach with the needed Javascript http://wiki.bitplan.com/index.php/MediaWiki:Common.js is already tricky and does not work in all environments out of the box. I got it working in some MediaWiki configurations but not in others. 

#70 Add Frontend with content negotiaton will later give us a lot more options. For the curators that do the main chore of entering events and event series it think the priorities are diffferent then for other stakeholders.

**Reply by MusaabKh** [reply]
I got the expand and collapse functionality of SMW working on forms. Is that okay or is HideShow a better alternative

**Reply by WolfgangFahl** [reply]
Can you share a link to an example or a screen shot and a link to the part of the page forms docs?

**Reply by WolfgangFahl** [reply]
Avoiding Javascript we'd have to maintain ourselves is a good idea!

**Reply by MusaabKh** [reply]
Here are some screenshots from the page 3DUI:
Collapsed:
![image](https://user-images.githubusercontent.com/26063969/108349056-5b2c5600-7204-11eb-81c4-781306fdaff8.png)
Expanded:
![image](https://user-images.githubusercontent.com/26063969/108349138-74cd9d80-7204-11eb-8c4c-dcfb208e14d3.png)
The link to the page:
https://confident.dbis.rwth-aachen.de/ormk/index.php?title=3DUI

The code snippet used: 
```
`{| class="mw-collapsible mw-collapsed wikitable"
! General Information
|-
| //anything here can be added which will be hidden
}`
```

**Reply by MusaabKh** [reply]
Another TestPage just for the Expand and Collapse Functionality:
https://confident.dbis.rwth-aachen.de/ormk/index.php?title=TestingInfoBox

**Reply by WolfgangFahl** [reply]
Great - let's see what the feedback will be.

**Reply by clange** [reply]
It's great to see so much creativity here in short time – special thanks to @MusaabKh and @tholzheim! 
Regarding @WolfgangFahl and your comment about "content negotiation", I could even imagine different entry forms for different _roles_. I don't want to make your job more complex, rather simpler. Feel free to move this to a separate thread.
* a simple form for the user story "as a non-participant, I want to enter information about an event I have heard about", 
* a slightly richer one for the user story "as a participant, I want to enter information about an event I have attended or am planning to attend"
* a fully featured one for "as an organizer (local organizer, proceedings chair, programme chair, …), I want to enter all details about my event".

Of course it should be easily possible to switch between them. I.e., to just have them as different views that everybody can choose from. Not that (heaven forbid!) OpenResearch links your actual role in each event to your user account and offers you restricted functionality based on that ;-)

**Comment by WolfgangFahl** [comment]
Here is a video on how not to implement things: https://youtu.be/aSyMV6wc8wI?list=TLGGqXaaNJWuzSoxODAyMjAyMQ. The forms are using lots of terms that are along the needs of the institution running the software instead of being aligned with the user needs. The forms are not self explaining and complicated. We'd love to be much simpler and self explaining close to what the stakeholders expect.

**Reply by WolfgangFahl** [reply]
If you want to have more fun like this read https://www.marktstammdatenregister.de/MaStRHilfe/files/regHilfen/Handbuch%20Aufforderung%20zur%20Datenkorrektur.pdf

**Comment by tholzheim** [comment]
I found a way to make the forms dynamically depended on the underlying properties by just providing the property for the field and querying the rest.
```
{{{field
          |{{#show: Property:Event id|?Property label}}
          |input type={{#show: Property:Event id|?Property inputType}}
          |placeholder={{#show: Property:Event id|?Property placeholder}}
          |{{#ifeq: {{#show: Property:Event id|?Property allowedValues }} |||values={{#show: Property:Event id|?Property allowedValues }} }}
          |{{#ifeq: {{#show: Property:Event id|?Property values from }} ||| values from {{#show: Property:Event id|?Property values from }} }}
          |{{#ifeq:{{#show: Property:Event id|?Property mandatory}}|true|mandatory}}
          |{{#ifeq:{{#show: Property:Event id|?Property primary key }}|true|unique}}
  }}}
```
>Note: This approach requires to adjust the type of some property properties and also add new property properties.

By defining all fields like this, a generator for forms can easily be implemented.

**Reply by MusaabKh** [reply]
I was already working on a form generator and this would work nicely for that.

**Reply by WolfgangFahl** [reply]
Could you please try out your results with "Template:Event" on your copies?

**Reply by tholzheim** [reply]
I improved this method by placing it in a [template](https://confident.dbis.rwth-aachen.de/orth/index.php?title=Template:FieldByProperty).
Now I can define a Form for Event as follows
```
{{{for template|Event|label=Event}}}
{| class="formtable BlueForm mw-collapsible" style="width:100%;"
!  colspan="2" class="BlueFormHeadline"|Basic information<nowiki>
   </nowiki>{{CreateFormRowByProperty|Property:Event id}}<nowiki>
   </nowiki>{{CreateFormRowByProperty|Property:Title}}<nowiki>
   </nowiki>{{CreateFormRowByProperty|Property:Event acronym}}<nowiki>
   </nowiki>{{CreateFormRowByProperty|Property:In event series}}<nowiki>
   </nowiki>{{CreateFormRowByProperty|Property:Ordinal}}<nowiki>
   </nowiki>{{CreateFormRowByProperty|Property:Event type}}
|}
```
The Form can be found [here](https://confident.dbis.rwth-aachen.de/orth/index.php?title=Form:Event).
This code uses a variety of different Templates that query the needed information from the properties.
![Untitled Diagram](https://user-images.githubusercontent.com/45004115/109513384-5342a080-7aa5-11eb-8eb1-17c92b61081e.png)
During the implementation of the templates I added the [Concept:Regexp](https://confident.dbis.rwth-aachen.de/orth/index.php?title=Concept:Regexp) which corresponds to https://www.mediawiki.org/wiki/Extension:Page_Forms/Input_types/Regexp it allows to define custom datatypes for fields. For example [Property:Ordinal](https://confident.dbis.rwth-aachen.de/orth/index.php?title=Property:Ordinal) only allows numbers, therefore, it uses [Regexp:NaturalNumber](https://confident.dbis.rwth-aachen.de/orth/index.php?title=Regexp:NaturalNumber) which restricts the input of the ordinal field to positive integers and shows an error message otherwise.
![Screenshot_2021-03-01 Edit Event Test - OPENRESEARCH th copy Wiki](https://user-images.githubusercontent.com/45004115/109514218-15924780-7aa6-11eb-8e13-7511c26a5410.png)
## Generating Forms
The use of templates to generate the fields also allows to generate the complete Form by querying the properties and formating the results by providing a template:
```
{{#ask: [[used for::Concept:Event]]|link=none|format=plainlist|template=CreateFormRowByProperty}}
```
### Complete Example
Here we generate a Form for Events with a categorization of the properties in
* mandatory
* recommended
* optional
>I am experimenting with different colors for the categories

by querying all information from the properties.
The Generated Form can be found [here](https://confident.dbis.rwth-aachen.de/orth/index.php?title=Form:TestGeneration)
#### Code
```
{{{for template|Event|label=Event}}}
----
{| class="formtable RedForm mw-collapsible" style="width:100%;"
!  colspan="2" class="BlueFormHeadline"|mandatory
{{#ask: [[used for::Concept:Event]][[Usage_obligation::mandatory]]|link=none|format=plainlist|template=CreateFormRowByProperty}}
|}
----
{| class="formtable BlueForm mw-collapsible mw-collapsed" style="width:100%;"
!  colspan="2" class="BlueFormHeadline"|Recommended
{{#ask: [[used for::Concept:Event]][[Usage_obligation::recommended]]|link=none|format=plainlist|template=CreateFormRowByProperty}}
|}
-----
{| class="formtable GreenForm mw-collapsible mw-collapsed" style="width:100%;"
!  colspan="2" class="BlueFormHeadline"|Optional
{{#ask: [[used for::Concept:Event]][[Usage_obligation::optional]]|link=none|format=plainlist|template=CreateFormRowByProperty}}
|}
----
```
#### Generated Form
![Screenshot_2021-03-01 Edit TestGeneration test - OPENRESEARCH th copy Wiki](https://user-images.githubusercontent.com/45004115/109515651-9bfb5900-7aa7-11eb-9dc9-7a277d366c48.png)

## Encountered Issues
* The properties of Event currently do not provide all information needed to generate the field. 
   >For this example I added all required information.
* Page Forms regexp does currently not wrk with all regexp (might be a version issue)
* Index and sortPos not defined for most of the properties
   * If defined it would allow sort the fields in the defined priority

**Reply by MusaabKh** [reply]
Another take on the collapsible forms. 

Sample images:
![Screenshot_20210305-105042__01](https://user-images.githubusercontent.com/26063969/110099067-2c37e780-7da1-11eb-9932-1ea936e74d1d.jpg)

Expanded:
![Screenshot_20210305-105048__01](https://user-images.githubusercontent.com/26063969/110099123-3eb22100-7da1-11eb-9cbd-122c800d9e67.jpg)

Sample link for the form: https://confident.dbis.rwth-aachen.de/ormk/index.php?title=3DUI_2020

**Comment by tholzheim** [comment]
Finally I found some better examples:
* Forms with collapsing fields: https://spore.fandom.com/wiki/Special:FormEdit/Fiction_character/test
![Screenshot_2021-03-01 FormEdit Fiction character test(1)](https://user-images.githubusercontent.com/45004115/109516400-6a36c200-7aa8-11eb-8bcd-2453849ab6c8.png)


* Forms with advanced styling: [example 1](http://edutechwiki.unige.ch/en/Special:FormEdit/Data_mining_and_learning_analytics_tools/test) and [example 2](http://edutechwiki.unige.ch/en/Special:FormEdit/Citizen_science_infrastructure/Test)

**Reply by WolfgangFahl** [reply]
Looks like we are making good progress. See https://www.openresearch.org/wiki/Heike_Prototypen
